File: app/src/main/java/guepardoapps/mycoins/customadapter/CoinListAdapter.java
package guepardoapps.mycoins.customadapter;

import java.util.ArrayList;
import java.util.Locale;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.View.OnClickListener;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.rey.material.app.Dialog;
import com.rey.material.app.ThemeManager;
import com.rey.material.widget.FloatingActionButton;

import guepardoapps.mycoins.CoinService;
import guepardoapps.mycoins.R;
import guepardoapps.mycoins.activities.ActivityCoinEdit;
import guepardoapps.mycoins.basic.dto.CoinDto;
import guepardoapps.mycoins.basic.enums.Currency;
import guepardoapps.mycoins.basic.utils.Logger;
import guepardoapps.mycoins.common.controller.NavigationController;

public class CoinListAdapter extends BaseAdapter {
    private static final String TAG = CoinListAdapter.class.getSimpleName();
    private Logger _logger;

    private Context _context;
    private NavigationController _navigationController;

    private ArrayList<CoinDto> _coinList;

    private static LayoutInflater _inflater = null;
    private CoinListAdapter _coinListAdapter;

    public CoinListAdapter(@NonNull Context context, @NonNull ArrayList<CoinDto> coinList) {
        _logger = new Logger(TAG, true);
        _logger.Debug("created...");

        _context = context;
        _navigationController = new NavigationController(_context);

        _coinList = coinList;

        _inflater = (LayoutInflater) _context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        _coinListAdapter = this;
    }

    @Override
    public int getCount() {
        return _coinList.size();
    }

    @Override
    public Object getItem(int position) {
        return position;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    private class Holder {
        ImageView _typeImage;
        TextView _type;
        TextView _amount;
        TextView _euroConversion;
        TextView _euroAmount;
        ImageView _euroAggregation;
        TextView _usDollarConversion;
        TextView _usDollarAmount;
        ImageView _usDollarAggregation;
        FloatingActionButton _btnDelete;
        FloatingActionButton _btnEdit;
    }

    @SuppressLint({"InflateParams", "ViewHolder"})
    @Override
    public View getView(final int index, View convertView, ViewGroup parent) {
        Holder holder = new Holder();
        View rowView = _inflater.inflate(R.layout.list_item, null);

        final CoinDto coin = _coinList.get(index);

        holder._typeImage = rowView.findViewById(R.id.itemTypeImage);
        holder._typeImage.setImageResource(coin.GetIcon());

        holder._type = rowView.findViewById(R.id.itemTypeText);
        holder._type.setText(coin.GetType());
        holder._amount = rowView.findViewById(R.id.itemAmount);
        holder._amount.setText(String.format(Locale.getDefault(), "%.6f", coin.GetAmount()));

        holder._euroConversion = rowView.findViewById(R.id.itemEuroConversion);
        holder._euroConversion.setText(coin.GetEuroConversionString());
        holder._euroAmount = rowView.findViewById(R.id.itemEuroAmount);
        holder._euroAmount.setText(coin.GetValueString(Currency.EUR));
        holder._euroAggregation = rowView.findViewById(R.id.itemEuroAggregation);
        if (coin.GetEuroAggregation() == CoinDto.Aggregation.Rise) {
            holder._euroAggregation.setImageResource(android.R.drawable.arrow_up_float);
        } else if (coin.GetEuroAggregation() == CoinDto.Aggregation.Fall) {
            holder._euroAggregation.setImageResource(android.R.drawable.arrow_down_float);
        } else {
            holder._euroAggregation.setImageResource(android.R.drawable.radiobutton_off_background);
        }

        holder._usDollarConversion = rowView.findViewById(R.id.itemUsDollarConversion);
        holder._usDollarConversion.setText(coin.GetUsDollarConversionString());
        holder._usDollarAmount = rowView.findViewById(R.id.itemUsDollarAmount);
        holder._usDollarAmount.setText(coin.GetValueString(Currency.USD));
        holder._usDollarAggregation = rowView.findViewById(R.id.itemUsDollarAggregation);
        if (coin.GetUsDollarAggregation() == CoinDto.Aggregation.Rise) {
            holder._usDollarAggregation.setImageResource(android.R.drawable.arrow_up_float);
        } else if (coin.GetUsDollarAggregation() == CoinDto.Aggregation.Fall) {
            holder._usDollarAggregation.setImageResource(android.R.drawable.arrow_down_float);
        } else {
            holder._usDollarAggregation.setImageResource(android.R.drawable.radiobutton_off_background);
        }

        holder._btnEdit = rowView.findViewById(R.id.btnEdit);
        holder._btnEdit.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _logger.Debug("_btnEdit onClick");

                coin.SetAction(CoinDto.Action.Update);

                Bundle details = new Bundle();
                details.putSerializable(CoinService.CoinIntent, coin);

                _navigationController.NavigateWithData(ActivityCoinEdit.class, details, false);
            }
        });

        holder._btnDelete = rowView.findViewById(R.id.btnDelete);
        holder._btnDelete.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _logger.Debug("_btnDelete onClick");

                boolean isLightTheme = ThemeManager.getInstance().getCurrentTheme() == 0;

                final Dialog dialog = new Dialog(_context);
                dialog
                        .title("Do you want to delete this note?")
                        .positiveAction("Yes")
                        .negativeAction("No")
                        .applyStyle(isLightTheme ? R.style.SimpleDialogLight : R.style.SimpleDialog)
                        .setCancelable(true);

                dialog.positiveActionClickListener(new OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        CoinService.getInstance().DeleteCoin(coin);
                        _coinList.remove(coin);
                        _coinListAdapter.notifyDataSetChanged();
                        dialog.dismiss();
                    }
                });

                dialog.negativeActionClickListener(new OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        dialog.dismiss();
                    }
                });

                dialog.show();
            }
        });

        return rowView;
    }
}

File: app/src/main/java/guepardoapps/mycoins/customadapter/CoinListAdapter.kt
package guepardoapps.mycoins.customadapter

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.afollestad.materialdialogs.MaterialDialog
import guepardoapps.mycoins.R
import guepardoapps.mycoins.activities.ActivityEdit
import guepardoapps.mycoins.common.Constants
import guepardoapps.mycoins.controller.NavigationController
import guepardoapps.mycoins.controller.SharedPreferenceController
import guepardoapps.mycoins.enums.Currency
import guepardoapps.mycoins.extensions.getSortFieldMethod
import guepardoapps.mycoins.extensions.getTrend
import guepardoapps.mycoins.models.Coin
import guepardoapps.mycoins.models.CoinAdapterHolder
import guepardoapps.mycoins.services.coin.CoinService
import kotlin.reflect.KCallable

@ExperimentalUnsignedTypes
internal class CoinListAdapter(private val context: Context, coinList: MutableList<Coin>, sortField: String, sortDirectionIsAsc: Boolean) : BaseAdapter() {

    private var coinAdapterList: List<CoinAdapterHolder> = listOf()

    private val inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    private val navigationController: NavigationController = NavigationController(context)

    init {
        updateDataSet(coinList, sortField, sortDirectionIsAsc)
    }

    fun updateDataSet(coinList: MutableList<Coin>, sortField: String, sortDirectionIsAsc: Boolean) {
        coinAdapterList = coinList.map { coin ->
            val currencyString = SharedPreferenceController(context).load(Constants.currency, Constants.currencyDefault)
            val currencyId = Currency.values().first { x -> x.text == currencyString }.id
            val currency = Currency.values()[currencyId]
            val coinConversion = CoinService.instance.getCoinConversion(coin.coinType)
            return@map CoinAdapterHolder(coin, currency, coinConversion)
        }

        coinAdapterList = if (sortDirectionIsAsc) {
            coinAdapterList.sortedWith(compareBy { coinAdapterHolder -> useCompareField(coinAdapterHolder, sortField) })
        } else {
            coinAdapterList.sortedWith(compareByDescending { coinAdapterHolder -> useCompareField(coinAdapterHolder, sortField) })
        }

        notifyDataSetChanged()
    }

    override fun getItem(position: Int): CoinAdapterHolder = coinAdapterList[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getCount(): Int = coinAdapterList.size

    @SuppressLint("SetTextI18n", "ViewHolder", "InflateParams")
    override fun getView(index: Int, convertView: View?, parentView: ViewGroup?): View {
        val rowView: View = inflater.inflate(R.layout.list_item, null)

        coinAdapterList[index].apply {
            coinImage = rowView.findViewById(R.id.coin_item_image)
            type = rowView.findViewById(R.id.coin_item_type)
            amount = rowView.findViewById(R.id.coin_item_amount)
            currencyImage = rowView.findViewById(R.id.coin_item_currency_image)
            currencyValue = rowView.findViewById(R.id.coin_item_currency_value)
            totalValue = rowView.findViewById(R.id.coin_item_total_value)
            trend = rowView.findViewById(R.id.coin_item_trend_icon)
            additionalInformation = rowView.findViewById(R.id.coin_item_additionalInformation)

            reload = rowView.findViewById(R.id.btnReload)
            edit = rowView.findViewById(R.id.btnEdit)
            delete = rowView.findViewById(R.id.btnDelete)

            coinImage.setImageResource(coin.coinType.iconId)

            type.text = type()
            amount.text = amountString()
            additionalInformation.text = additionalInformation()
            totalValue.text = totalValueString()
            currencyValue.text = currencyValueString()

            currencyImage.setImageResource(if (currency == Currency.EUR) R.mipmap.euro else if (currency == Currency.USD) R.mipmap.dollar else R.drawable.dummy)

            trend.setImageResource(CoinService.instance.getCoinTrend(coin.coinType).getTrend().id)

            reload.setOnClickListener {
                CoinService.instance.loadCoinConversion(coin.coinType)
                CoinService.instance.loadCoinTrend(coin.coinType)
            }

            edit.setOnClickListener {
                val bundle = Bundle()
                bundle.putString(Constants.bundleDataId, coin.id)
                navigationController.navigateWithData(ActivityEdit::class.java, bundle, false)
            }

            delete.setOnClickListener {
                MaterialDialog(context).show {
                    title(text = context.getString(R.string.delete))
                    message(text = "${context.getString(R.string.delete)} ${coin.coinType.type}?")
                    positiveButton(text = context.getString(R.string.yes)) { CoinService.instance.deleteCoin(coin) }
                    negativeButton(text = context.getString(R.string.no))
                }
            }
        }

        return rowView
    }

    @Suppress("UNCHECKED_CAST")
    private fun useCompareField(coinAdapterHolder: CoinAdapterHolder, sortField: String): Comparable<*> {
        val kPair = coinAdapterHolder.getSortFieldMethod(sortField)
        return if (kPair.first != null) {
            when (kPair.second) {
                Double::class -> (kPair.first as KCallable<Double>).call(coinAdapterHolder)
                else -> (kPair.first as KCallable<String>).call(coinAdapterHolder)
            }
        } else {
            coinAdapterHolder.type()
        }
    }
}